driver: gpio: add low-voltage (1.8v) level detection in npcx series.

MulinChao opened the pull request:
Part of GPIO pads in npcx series support low-voltage (1.8V) level
detection. In order to introduce this feature, this CL adds a new
NPCX-specific controller property, lvol_io_pads, in devicetree file.
For example, here is devicetree fragment which turn on low-voltage
support of i2c1_0 port.
 

    /{  
       def_lvol_io_list {
          compatible = "nuvoton,npcx-lvolctrl-def";
          lvol_io_pads = <&lvol_io90 /* I2C1_SCL0 1.8V support */
                          &lvol_io87>; /* I2C1_SDA0 1,8V support */
     };
  };

Then these pads will turn on 1.8V level detection during initialization.

MulinChao committed:
driver: gpio: add 1p8v level detection support in npcx series.

Part of GPIO pads in npcx series support low-voltage (1.8V) level
detection. In order to introduce this feature, this CL adds a new
NPCX-specific controller property, lvol_io_pads, in devicetree file.
For example, here is devicetree fragment which turn on low-voltage
support of i2c1_0 port.

/ {
      def_lvol_io_list {
          compatible = "nuvoton,npcx-lvolctrl-def";
          lvol_io_pads = <&lvol_io90   /* I2C1_SCL0 1.8V support */
                          &lvol_io87>; /* I2C1_SDA0 1,8V support */
     };
  };

Then these pads will turn on 1.8V level detection during initialization.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

pabigot commented on include/drivers/gpio.h:
```diff
@@ -201,6 +201,13 @@ extern "C" {
+ * @note Drivers that do not support 1.8V level detection feature should ignore
+ * this flag rather than rejecting the configuration with -ENOTSUP.
+ */
+#define GPIO_INT_1P8V_SEL              (1U << 20)
```
I don't think this is the right way to solve this.  It opens up to having more flags for 1.5V, 2.5V, 3.0V, 3.3V, 5.0V, ....

If you don't need to change the voltage configuration at runtime I'd rather see a NPCX-specific controller property in devicetree.  [This](https://github.com/torvalds/linux/blob/5c8fe583cce542aa0b84adc939ce85293de36e5e/Documentation/devicetree/bindings/arm/tegra/nvidia%2Ctegra20-pmc.yaml#L283) would be one approach: a bitmask property that indicates which pins are to be configured for which voltage.  If you can find others that have precedence in Linux they could be considered too.

If you need to change at runtime (which seems unlikely) I think you'd have to have driver-specific API extensions.

MulinChao replied:
Got it. Thanks for your suggestion. I will remove this change in gpio.h. Study adding new property now.

MulinChao replied:
Done.

pabigot reviewed (changes requested):
I don't think we need to (nor should) change the GPIO flags for this.

pabigot commented on boards/arm/npcx7m6fb_evb/npcx7m6fb_evb.dts:
```diff
@@ -33,6 +33,13 @@
 		};
 	};
+
+	def_lvol_io_list {
```
Does it make sense to pull some of this out to a separate commit after 1.8V support is added in a first commit?

I think this node should be introduced in the main commit with an empty `lvol_io_pads`.  Then add the I2C1 pad definitions to it in a second commit.

That would serve as an example of how to use this feature, which can be consulted by people who want to add 1.8V support for other functions without getting lost in all the other details required to support it generally.

MulinChao replied:
Good idea. Will separate it into two CLs later.

MulinChao replied:
Done.

MulinChao committed:
boards: arm: npcx7m6fb_evb: An example for low-voltage feature.

This CL provided an example of how turns on the low-voltage level
detection feature in npcx series. It demonstrates enabling low-voltage
level detection of I2C1_0 SCL/SDA io-pads if the power rail of their PUs
is 1.8V.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

pabigot commented on boards/arm/npcx7m6fb_evb/npcx7m6fb_evb.dts:
```diff
@@ -36,7 +36,8 @@
 	def_lvol_io_list {
 		compatible = "nuvoton,npcx-lvolctrl-def";
-		lvol_io_pads = <>; /* Put low-voltage io pads here */
+		lvol_io_pads = <&lvol_io90   /* I2C1_SCL0 1.8V support */
```
Is this just a demonstration, or is this something the board actually needs?  IOW could somebody put 3.3V devices on this bus, then have to remove this from the board devicetree to make things work?

MulinChao replied:
@pabigot :
It depends on what customers need. There're no default PU resistances on i2c's pads of npcx7m6fb's EVB. (Please refer bottom right corner of the [picture](https://docs.zephyrproject.org/latest/_images/npcx7m6fb_evb.png)). The PU resistance value influences i2c's bus frequency a lot, hence we don't put them on i2c pads of EVB by default. 

There're two cases for our i2c pads:
1. The bus voltage of i2c device is 3.3V. Since the origin detection level of i2c's pads is 3.3V (ie. Vih = 3.3V*0.7 and Vil = 3.3V *0.3), the customers needn't put these pads into "lvol_io_pads" array. They can connect 3.3 PU via the other circuit or our PU connectors which provide 3.3 PU on EVB.
2. The bus voltage of i2c device is 1.8V. The customers need to put these pads into "lvol_io_pads" array and their detection level changes to 1.8V. (ie. Vih = 1.8V * 0.7 and Vil = 1.8V  * 0.3). They can connect 1.8 PU via the other circuit or our PU connectors which provide 1.8V PU on EVB.

pabigot replied:
So it sounds like the default configuration is 3.3V, and the board does not require that it be anything else.  In that case I don't think the board devicetree should be changing it to a non-default configuration as an example.  The instructions for changing PU levels should be somewhere other than the operational part of the devicetree.

I don't think it's a good idea, but since the board is not available to casual users who could be asking questions on Zephyr support forums, if you want to change the behavior as an example I wouldn't block it.  You should probably add a comment though saying that it's not necessarily correct.  Or put the properties inside such a comment.

MulinChao replied:
Thanks for the suggestion. I will put this example into a comment before this property.
